### PR TITLE
Dash / Tunnel Docs

### PR DESCRIPTION
**Pull request by r4vi**
Updates documentation for how to use the new `_tunnel_` endpoint. This walks through a hello-world example with required changes to use with our special endpoint.

### CODE REVIEW

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+if __name__ == '__main__':
+    app.run_server(host='0.0.0.0')
+
+```
+
+There are two important things to note here, where this code differs from the
+plotly dash example:
+
+The `Dash` class must be instanciated with a `url_base_pathname`. This should
```
Comment (davidread) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
The `Dash` class must be instantiated with a `url_base_pathname`. This should
```

Comment (r4vi) [reply 16-Access-Local-Ports-in-Jupyter.Rmd]: good spot :pray:

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 2]
```diff
@@ -0,0 +1,137 @@
+```{r global_options, include=FALSE}
+{knitr::opts_chunk$set(eval = FALSE)}
+```
+# Access Local Ports in Jupyter
+If you are developing a web application using Jupyter you will probably want to
+preview that webapp before you deploy it. The platform provides a special set of
+endpoints under `/_tunnnel_/<portNumber>/` that route to exposed HTTP ports on
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
endpoints under `/_tunnel_/<portNumber>/` that route to exposed HTTP ports on
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+the localhost. There are a few things you need to keep in mind:
+
+* `<portNumber>` can only be one from a limited rage. By default you can only route
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
* `<portNumber>` can only be one from a limited range. By default, you can only route
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  to ports 8050 and 4040-4050 inclusive. You should make your webapp listen on
+  one of those.
+* The `/_tunnel_/<portNumber>/` endpoint will only tunnel to services bound to a
+  non public IP address. By default many web frameworks bind to host `127.0.0.1`
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
  non--public IP address. By default, many web frameworks bind to host `127.0.0.1`.
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  you will need to change this to `0.0.0.0` or the tunnel won't work. This is to
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
  You will need to change this to `0.0.0.0` or the tunnel won't work. This is to
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+pip install --user dash==0.39.0  # The core dash backend
+pip install --user dash-daq==0.1.0  # DAQ components (newly open-sourced!)
+```
+
+The demo code in this document is known to work with these versions but you can
+install any version you like for your own code.
+
+![](images/dash/dash_install_deps.gif)
+
+If you are planning on turning this into a project you'll need to manage your
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
If you are planning on turning this into a project, you'll need to manage your
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+dependencies by adding these to either `requirements.txt` & `pip` or
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
dependencies by adding these to either `requirements.txt` and `pip` or
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+`environment.yaml` & `conda`.
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
`environment.yaml` and `conda`.
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+### Example code
+Save the dash hello world to a new python file called `app.py`:
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
Save the Dash hello world app to a new python file called `app.py`:
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        }
+    }
+
+if __name__ == '__main__':
+    app.run_server(host='0.0.0.0')
+
+```
+
+There are two important things to note here, where this code differs from the
+plotly dash example:
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
Plotly Dash example:
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+The `Dash` class must be instantiated with a `url_base_pathname`. This should
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
1. The `Dash` class must be instantiated with a `url_base_pathname`. This should
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+always be `/_tunnel_/8050/` e.g.
+`app = dash.Dash('app', server=server, url_base_pathname='/_tunnel_/8050/')`.
+
+When you run the server it must be bound to a `0.0.0.0` e.g. `app.run_server(host='0.0.0.0')`
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
2. When you run the server, it must be bound to `0.0.0.0`, e.g., `app.run_server(host='0.0.0.0')`.
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+ * Environment: production
+   WARNING: Do not use the development server in a production environment.
+   Use a production WSGI server instead.
+ * Debug mode: off
+ * Running on http://0.0.0.0:8050/ (Press CTRL+C to quit)
+```
+
+### Access via `_tunnel_` url
+
+Copy your `jupyter` URL and append `/_tunnel_/8050/` to access your running dash
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
Copy your `jupyter` URL and append `/_tunnel_/8050/` to access your running Dash
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+app, if you're `jupyter` url is: `https://r4vi-jupyter-lab.tools.alpha.mojanalytics.xyz/lab?`
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
app. If you're `jupyter` URL is `https://r4vi-jupyter-lab.tools.alpha.mojanalytics.xyz/lab?`,
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 15]
```diff
@@ -0,0 +1,137 @@
+```{r global_options, include=FALSE}
+{knitr::opts_chunk$set(eval = FALSE)}
+```
+# Access Local Ports in Jupyter
+If you are developing a web application using Jupyter you will probably want to
+preview that webapp before you deploy it. The platform provides a special set of
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
preview that web app before you deploy it. The Analytical Platform provides a special set of
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+endpoints under `/_tunnnel_/<portNumber>/` that route to exposed HTTP ports on
+the localhost. There are a few things you need to keep in mind:
+
+* `<portNumber>` can only be one from a limited rage. By default you can only route
+  to ports 8050 and 4040-4050 inclusive. You should make your webapp listen on
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
  to ports 8050 and 4040--4050 inclusive. You should make your web app listen on
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+  one of those.
+* The `/_tunnel_/<portNumber>/` endpoint will only tunnel to services bound to a
+  non public IP address. By default many web frameworks bind to host `127.0.0.1`
+  you will need to change this to `0.0.0.0` or the tunnel won't work. This is to
+  prevent inadvertently exposing webapps to the tunnel.
+
+## Running Plotly [Dash] apps
+
+### Install dependencies
+In the terminal install the dash dependencies:
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
In the terminal, install the Dash dependencies:
```

File: 16-Access-Local-Ports-in-Jupyter.Rmd [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+yovyan@jupyter-lab-r4vi-jupyter-7cb4bb58cf-6hngf:~$ python3 app.py
+ * Serving Flask app "app" (lazy loading)
+ * Environment: production
+   WARNING: Do not use the development server in a production environment.
+   Use a production WSGI server instead.
+ * Debug mode: off
+ * Running on http://0.0.0.0:8050/ (Press CTRL+C to quit)
+```
+
+### Access via `_tunnel_` url
```
Comment (calumabarnett) [inline 16-Access-Local-Ports-in-Jupyter.Rmd]:
```suggestion
### Access via `_tunnel_` URL
```

**Comment by r4vi** [comment]
@calumabarnett thanks for the suggestions!

**Pull request merged**